File: ios/RTSDA/ViewModels/EventsViewModel.swift
import Foundation
import FirebaseFirestore
import Combine

enum EventsError: LocalizedError {
    case fetchFailed(String)
    case addFailed(String)
    case updateFailed(String)
    case deleteFailed(String)
    case networkError(String)
    
    var errorDescription: String? {
        switch self {
        case .fetchFailed(let message):
            return "Failed to fetch events: \(message)"
        case .addFailed(let message):
            return "Failed to add event: \(message)"
        case .updateFailed(let message):
            return "Failed to update event: \(message)"
        case .deleteFailed(let message):
            return "Failed to delete event: \(message)"
        case .networkError(let message):
            return "Network error: \(message)"
        }
    }
}

@MainActor
class EventsViewModel: ObservableObject {
    @Published var events: [Event] = []
    @Published var isLoading = false
    @Published var error: EventsError?
    
    private let db = Firestore.firestore()
    private var listenerRegistration: ListenerRegistration?
    
    init() {
        setupEventsListener()
    }
    
    deinit {
        listenerRegistration?.remove()
    }
    
    private func setupEventsListener() {
        isLoading = true
        error = nil
        
        listenerRegistration = db.collection("events")
            .whereField("isPublished", isEqualTo: true)
            .order(by: "startDate")
            .order(by: FieldPath.documentID())
            .addSnapshotListener { [weak self] (querySnapshot: QuerySnapshot?, error: Error?) in
                guard let self = self else { return }
                
                if let error = error {
                    self.error = .fetchFailed(error.localizedDescription)
                    self.isLoading = false
                    return
                }
                
                do {
                    self.events = try querySnapshot?.documents
                        .compactMap { document -> Event? in
                            do {
                                if let event = try Event.fromDocument(document) {
                                    // Double check isPublished in case of race conditions
                                    return event.isPublished ? event : nil
                                }
                                return nil
                            } catch {
                                print("Error parsing event document: \(error)")
                                return nil
                            }
                        }
                        .sorted { $0.startDate < $1.startDate }
                        ?? []
                    self.isLoading = false
                } catch {
                    self.error = .fetchFailed("Failed to parse events: \(error.localizedDescription)")
                    self.isLoading = false
                }
            }
    }
    
    func refresh() {
        listenerRegistration?.remove()
        setupEventsListener()
    }
    
    func addEvent(_ event: Event) async {
        isLoading = true
        error = nil
        
        do {
            _ = try await db.collection("events")
                .addDocument(data: event.toDocument())
            
            isLoading = false
        } catch {
            self.error = .addFailed(error.localizedDescription)
            isLoading = false
        }
    }
    
    func updateEvent(_ event: Event) async {
        isLoading = true
        error = nil
        
        do {
            try await db.collection("events")
                .document(event.id)
                .setData(event.toDocument(), merge: true)
            
            isLoading = false
        } catch {
            self.error = .updateFailed(error.localizedDescription)
            isLoading = false
        }
    }
    
    func deleteEvent(_ event: Event) async {
        isLoading = true
        error = nil
        
        do {
            try await db.collection("events")
                .document(event.id)
                .delete()
            
            isLoading = false
        } catch {
            self.error = .deleteFailed(error.localizedDescription)
            isLoading = false
        }
    }
}

File: android/app/src/main/java/com/rtsda/appr/ui/viewmodels/EventsViewModel.kt
package com.rtsda.appr.ui.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import com.rtsda.appr.data.model.CalendarEvent
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import timber.log.Timber
import javax.inject.Inject

sealed class EventsError(val message: String) {
    class FetchFailed(message: String) : EventsError("Failed to fetch events: $message")
    class AddFailed(message: String) : EventsError("Failed to add event: $message")
    class UpdateFailed(message: String) : EventsError("Failed to update event: $message")
    class DeleteFailed(message: String) : EventsError("Failed to delete event: $message")
    class NetworkError(message: String) : EventsError("Network error: $message")
}

data class EventsUiState(
    val events: List<CalendarEvent> = emptyList(),
    val isLoading: Boolean = false,
    val error: EventsError? = null
)

@HiltViewModel
class EventsViewModel @Inject constructor(
    private val db: FirebaseFirestore
) : ViewModel() {

    private val _uiState = MutableStateFlow(EventsUiState(isLoading = true))
    val uiState: StateFlow<EventsUiState> = _uiState.asStateFlow()

    private var eventsListener: ListenerRegistration? = null

    init {
        setupEventsListener()
    }

    private fun setupEventsListener() {
        eventsListener?.remove()
        
        _uiState.update { it.copy(isLoading = true, error = null) }
        
        eventsListener = db.collection("events")
            .whereEqualTo("isPublished", true)
            .orderBy("startDate", Query.Direction.ASCENDING)
            .orderBy("id", Query.Direction.ASCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    Timber.e(error, "Error fetching events")
                    _uiState.update { 
                        it.copy(
                            isLoading = false,
                            error = EventsError.FetchFailed(error.message ?: "Unknown error")
                        )
                    }
                    return@addSnapshotListener
                }

                snapshot?.let { querySnapshot ->
                    try {
                        val events = querySnapshot.documents
                            .mapNotNull { doc ->
                                try {
                                    CalendarEvent.fromDocument(doc)?.takeIf { it.isPublished && !it.isDeleted }
                                } catch (e: Exception) {
                                    Timber.e(e, "Error parsing event ${doc.id}")
                                    null
                                }
                            }
                            .sortedBy { it.startDate.seconds }
                        
                        _uiState.update { 
                            it.copy(
                                events = events,
                                isLoading = false,
                                error = null
                            )
                        }
                    } catch (e: Exception) {
                        Timber.e(e, "Error processing events")
                        _uiState.update { 
                            it.copy(
                                isLoading = false,
                                error = EventsError.FetchFailed(e.message ?: "Unknown error")
                            )
                        }
                    }
                }
            }
    }

    fun loadEvents() {
        setupEventsListener()
    }

    fun refresh() {
        setupEventsListener()
    }

    suspend fun addEvent(event: CalendarEvent) {
        _uiState.update { it.copy(isLoading = true, error = null) }
        
        try {
            db.collection("events")
                .add(event.toMap())
                .await()
            
            _uiState.update { it.copy(isLoading = false) }
        } catch (e: Exception) {
            Timber.e(e, "Error adding event")
            _uiState.update { 
                it.copy(
                    isLoading = false,
                    error = EventsError.AddFailed(e.message ?: "Unknown error")
                )
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        eventsListener?.remove()
    }
}
